src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+      return op;
+    }
+
+    const activeSelections = Object.entries(typeFragments).reduce(
+      (state, [key, value]) => ({
+        ...state,
+        [key]: value.filter(s => activeOperations.has(s.key)),
+      }),
+      typeFragments
+    );
+
+    return {
+      ...op,
+      query: addFragmentsToQuery({
+        schema,
+        typeFragments: activeSelections,
+        userFragments: userFragments,
+        query: op.query,
+      }),
+    };
+  };
+
+  /** Handle query and extract fragments. */
+  const handleIncomingQuery = ({ key, operationName, query }: Operation) => {
+    activeOperations.add(key);
+
+    if (operationName !== 'query' || parsedOperations.has(key)) {
+      return;
+    }
+
+    parsedOperations.add(key);
+
+    const {
+      fragments: newFragments,
+      selections: newSelections,
+    } = extractSelectionsFromQuery({
+      schema,
+      query,
+    });
+
+    userFragments = newFragments.reduce(
+      (state, fragment) => ({
+        ...state,
+        [fragment.name.value]: fragment,
+      }),
+      userFragments
+    );
+
+    typeFragments = newSelections.reduce((state, { selections, type }) => {
+      const current = state[type] || [];
+      const entry: TypeFragment = {
+        key,
+        fragment: {
+          kind: 'FragmentDefinition',
+          typeCondition: {
+            kind: 'NamedType',
+            name: {
+              kind: 'Name',
+              value: type,
+            },
+          },
+          name: {
+            kind: 'Name',
+            value: `${type}_PopulateFragment_${current.length}`,
+          },
+          selectionSet: selections,
+        },
+        type,
+      };
+      return {
+        ...state,
+        [type]: [...current, entry],
+      };
+    }, typeFragments);
+  };
+
+  const handleIncomingTeardown = ({ key, operationName }: Operation) => {
+    if (operationName !== 'teardown') {
+      return;
+    }
+
+    activeOperations.delete(key);
+  };
+
+  return ops$ => {
+    return pipe(
+      ops$,
+      tap(handleIncomingQuery),
+      tap(handleIncomingTeardown),
+      map(handleIncomingMutation),
+      forward
+    );
+  };
+};
+
+type UserFragmentMap<T extends string = string> = Record<
+  T,
+  FragmentDefinitionNode
+>;
+
+type TypeFragmentMap<T extends string = string> = Record<T, TypeFragment[]>;
+
+interface TypeFragment {
+  /** Operation key where selection set is being used. */
+  key: number;
+  /** Selection set. */
+  fragment: FragmentDefinitionNode;
+  /** Type of selection. */
+  type: string;
+}
+
+interface MakeFragmentsFromQueryArg {
+  schema: GraphQLSchema;
+  query: DocumentNode;
+}
+
+/** Gets typed selection sets and fragments from query */
+export const extractSelectionsFromQuery = ({
+  schema,
+  query,
+}: MakeFragmentsFromQueryArg) => {
+  let selections: { selections: SelectionSetNode; type: string }[] = [];
+  let fragments: FragmentDefinitionNode[] = [];
+  const typeInfo = new TypeInfo(schema);
```

Comment:
Seems like this instance should maybe be preserved and only be instantiated once?

Reply:
TypeInfo is stateful and is used while traversing the schema alongside a given query/mutation. Not too sure how else we would go about this - any examples?

Reply:
Ok, didn’t realise this was stateful 👍 the name didn’t really imply it 😂